chain/chain/src/chain.rs

```diff
@@ -525,14 +528,21 @@ impl Chain {
             )
             .into());
         }
+        let prev_epoch_id = self.get_block_header(&head.prev_block_hash)?.epoch_id();
+        let epoch_change = prev_epoch_id != &head.epoch_id;
+        let fork_tail = if epoch_change { gc_stop_height } else { self.store.fork_tail()? };
```

Comment:
It seems you need to call `update_fork_tail` here. Otherwise it's possible that clear_forks_data() executes well, gc_blocks_remaining == 0 and you will return with no fork_tail updated. On the next height there will be no epoch_change.